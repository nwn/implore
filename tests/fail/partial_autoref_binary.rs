use implore::impl_op;

#[derive(Copy, Clone, PartialEq, Eq, Debug)]
struct X;

#[impl_op(autoref)]
fn sub(lhs: X, _rhs: &X) -> X {
    lhs
}

fn partial_subtraction() {
    assert_eq!(X, X - X);
    assert_eq!(X, X - &X);
    assert_eq!(X, &X - X);
    assert_eq!(X, &X - &X);
}

#[impl_op(autoref)]
fn mul(_lhs: &X, rhs: X) -> X {
    rhs
}

fn partial_multiplication() {
    assert_eq!(X, X * X);
    assert_eq!(X, X * &X);
    assert_eq!(X, &X * X);
    assert_eq!(X, &X * &X);
}

fn main() {}
