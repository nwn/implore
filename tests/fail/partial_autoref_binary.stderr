error[E0369]: cannot subtract `X` from `&X`
  --> $DIR/partial_autoref_binary.rs:14:22
   |
14 |     assert_eq!(X, &X - X);
   |                   -- ^ - X
   |                   |
   |                   &X
   |
help: `-` can be used on `X`, you can dereference `&X`
   |
14 |     assert_eq!(X, *&X - X);
   |                   ^

error[E0369]: cannot subtract `&X` from `&X`
  --> $DIR/partial_autoref_binary.rs:15:22
   |
15 |     assert_eq!(X, &X - &X);
   |                   -- ^ -- &X
   |                   |
   |                   &X
   |
help: `-` can be used on `X`, you can dereference `&X`
   |
15 |     assert_eq!(X, *&X - &X);
   |                   ^

error[E0308]: mismatched types
  --> $DIR/partial_autoref_binary.rs:25:23
   |
25 |     assert_eq!(X, X * &X);
   |                       ^^
   |                       |
   |                       expected struct `X`, found `&X`
   |                       help: consider removing the borrow: `X`

error[E0308]: mismatched types
  --> $DIR/partial_autoref_binary.rs:27:24
   |
27 |     assert_eq!(X, &X * &X);
   |                        ^^
   |                        |
   |                        expected struct `X`, found `&X`
   |                        help: consider removing the borrow: `X`
